[ClangImporter] Import Swift 3 and 4 names for enumerators.

jrose-apple opened the pull request:
Also lays the groundwork for rdar://problem/16513537, which is about being able to find an enum by its original top-level name so that we can show a diagnostic for that. I'll file a public bug about that later.

rdar://problem/31893305

jrose-apple commented:
@swift-ci Please test

milseman commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -345,6 +345,23 @@ static bool isNSDictionaryMethod(const clang::ObjCMethodDecl *MD,
+    ClangImporter::Implementation &impl, const clang::NamedDecl *decl,
+    llvm::function_ref<void(ImportedName, ImportNameVersion)> action) {
+  using ImportNameKey = std::pair<DeclName, EffectiveClangContext>;
+  SmallVector<ImportNameKey, 8> seenNames;
```
Perhaps SmallSet? Yeah, std::set is not ideal when it grows outside the static size, but at least it has reasonable complexity when very large.

If you're opposed to SmallSet, then DeclName is pointer-like. For EffectiveClangContext, do you want the unresolved name here, or the Clang decl, or what are you expecting? An unresolved name might compare unequal to the same name but resolved to a Clang decl, for example.

jrose-apple replied:
EffectiveClangContext isn't orderable, so I can't use a normal set. But with N < 8 ImportNameVersions and most of them being the same I'm not too worried about N^2 behavior here.

I think I'm going to not worry about UnresolvedNames for now.

milseman commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -345,6 +345,23 @@ static bool isNSDictionaryMethod(const clang::ObjCMethodDecl *MD,
   return true;
 }
 
+static void forEachDistinctName(
```
Could you provide a comment on what "distinct" means and any assumptions? Swift name is distinct for Clang name? Maybe each name for each version for the imported decl? Or can a given version have many Swift names?

jrose-apple replied:
Hm. It's each distinct name and context to import the declaration, at most one per Swift version. I'm happy to put that in a doc comment, but do you have a better name for the function itself?

milseman replied:
Function name seems fine to me. Would like to understand it by reading a comment rather than reading the body.

milseman commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -345,6 +345,23 @@ static bool isNSDictionaryMethod(const clang::ObjCMethodDecl *MD,
+                                      [&](ImportNameVersion nameVersion) {
+    // Check to see if the name is different.
+    ImportedName newName = impl.importFullName(decl, nameVersion);
+    ImportNameKey key(newName, newName.getEffectiveContext());
```
importFullName will try the NameImporter cache, which is keyed off of {clang Decl, version} and results in a ImportedName. Does your notion of uniqueness differ?

jrose-apple replied:
Yes, Swift 2 and Swift 3 might have the same imported name, in which case we don't actually want to import the declaration twice.

milseman commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -2610,23 +2629,55 @@ namespace {
+          return underlyingTy->getAsTagDecl() == canonicalClangDecl;
+        }
+        case EffectiveClangContext::UnresolvedContext:
+          llvm_unreachable("can't import into an unresolved context");
```
Refresh my memory (ideally with a comment). Why couldn't we ever hit this?

jrose-apple replied:
I think the only way we could get here is if someone renamed an enumerator onto *another* decl, which won't work. But we shouldn't crash. I'll change it to just return false.

milseman commented on lib/ClangImporter/SwiftLookupTable.h:
```diff
@@ -133,6 +133,21 @@ class EffectiveClangContext {
+  /// Compares two EffectiveClangContexts without resolving unresolved names.
+  friend bool operator==(const EffectiveClangContext &lhs,
+                         const EffectiveClangContext &rhs) {
+    if (lhs.getKind() != rhs.getKind())
```
As mentioned earlier, this will yield a false negative when one is resolved and not the other. Not sure if that's a scenario which is important to you. Might be worth having a function with a name rather than ==.

jrose-apple replied:
Fair enough.

milseman reviewed (commented):
Functionality looks good to me. Have some questions.

jrose-apple commented:
Thanks for reviewing!

jrose-apple committed:
[ClangImporter] Import Swift 3 and 4 names for enumerators.

Also lays the groundwork for rdar://problem/16513537, which is about
being able to find an enum by its original top-level name so that we
can show a diagnostic for that. I'll file a public bug about that
later.

rdar://problem/31893305

jrose-apple commented:
@swift-ci Please smoke test
